users/rLapz/server.c

```diff
@@ -1,219 +1,494 @@
+
+			return -1;
+		}
+
+		s->fds.pfds = tmp_srv;
+		s->client   = tmp_client;
+	}
+
+	int i = s->fds.fd_count;
+
+	s->fds.pfds[i].fd     = new_fd;
+	s->fds.pfds[i].events = POLLIN;
+
+	s->client[i].fd   = new_fd;
+	s->client[i].addr = *(s->addr); /* copy the value */
+	s->client[i].port = htons(get_port((struct sockaddr *)&(s->client[i].addr)));
+
+	get_addr_str(s->client[i].addr_str, (struct sockaddr *)&(s->client[i].addr));
+
+
+	printf("server: new connection from [%s (%d)] on socket %d\n\n",
+			s->client[i].addr_str, s->client[i].port, new_fd
+	);
+
+	(s->fds.fd_count)++;
+
+	return s->fds.fd_count;
+}
+
+
+static void
+server_poll(struct server *s)
 {
-	short int is_error		= FALSE;
-	int socket_option		= 1;
-	int server_fd			= 0;
-	int client_fd			= 0;
-	struct sockaddr_in server	= {0};
-	struct sockaddr_in client	= {0};
-	socklen_t client_len		= 0;
-
-	/* Create socket */
-	server_fd = socket(AF_INET, SOCK_STREAM, 0);
-	if (server_fd < 0) {
-		perror("Create socket");
-		is_error = TRUE;
-		goto cleanup;
-	}
-
-	/* Socket option
-	 * Reuse IP address */
-	if (setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR,
-			&socket_option, sizeof(socket_option)) < 0) {
-		perror("Reuse address");
-		is_error = TRUE;
-		goto cleanup;
-	}
-
-	/* TCP configuration */
-	server.sin_family = AF_INET;
-	server.sin_addr.s_addr = inet_addr(addr);
-	server.sin_port = htons(port);
-
-	/* binding server address */
-	if (bind(server_fd, (struct sockaddr *)&server, sizeof(server)) < 0) {
-		perror("Bind socket");
-		is_error = TRUE;
-		goto cleanup;
-	}
-
-	/* listening... */
-	if (listen(server_fd, 3) < 0) {
-		perror("Listening");
-		is_error = TRUE;
-		goto cleanup;
-	}
-	printf("Listening: [%s:%d]...\n", addr, port);
-
-	/* get file properties from client */
-	while (1) {
-		client_len = sizeof(struct sockaddr_in);
-		puts("Waiting for incoming connection...");
-		client_fd = accept(server_fd, (struct sockaddr *)&client, &client_len);
-		if (client_fd < 0) {
-			perror("Accepting connection");
+	int       new_fd;
+	struct    sockaddr_storage addr;
+	socklen_t addr_len = sizeof(addr);
+
+	s->addr           = &addr;
+	s->fds.poll_count = poll(s->fds.pfds, s->fds.fd_count, -1);
+
+	if (s->fds.poll_count == -1)
+		perror("server: server_poll(): poll");
+
+	for (int i = 0; i < s->fds.fd_count; i++) {
+		if (s->fds.pfds[i].revents != POLLIN)
 			continue;
+
+		if (s->fds.pfds[i].fd == s->listener) {
+			new_fd = accept(s->listener,
+					(struct sockaddr *)&addr, &addr_len);
+
+			if (new_fd < 0) {
+				perror("server: server_poll(): accept");
+
+				continue;
+			}
+
+			if (add_to_pfds(s, new_fd) < 0) {
+				clean_up(s);
+				die("server: server_poll()");
+			}
+
 		} else {
-			puts("Connection accepted");
-			file_handler(&client_fd);
+			/* TODO: Multithreading support (file I/O) */
+	
+			file_io(s, i);
+	
+			close(s->fds.pfds[i].fd);
+	
+			printf("server: client on socket %d has closed\n",
+					s->fds.pfds[i].fd
+			);
+	
+			delete_from_pfds(s, i);
```

Comment:
Not surprised why your server is hang, you wait for `file_io()` to finish every single connection. You need to context switch them, for instance.
- poll()
- accept() = get client 1
- poll()
- recv(client 1)
- write(file 1) /* client 1 hasn't finished */
- poll()
- accept() = get client 2
- poll()
- recv(client 2)
- write(file 2) /* client 2 has finished */
- close(client 2)
- poll()
- recv(client 1) /* continue receiving from client 1 */
- write(client 1)
- .... and so on...

Reply:
OK, Thanks for the helpful advice!